--- a/net/minecraft/client/resources/model/DelegateBakedModel.java
+++ b/net/minecraft/client/resources/model/DelegateBakedModel.java
@@ -25,8 +_,18 @@
     }
 
     @Override
+    public List<BakedQuad> getQuads(@Nullable BlockState state, @Nullable Direction side, RandomSource rand, net.minecraftforge.client.model.data.ModelData extraData, @Nullable net.minecraft.client.renderer.RenderType renderType) {
+        return this.parent.getQuads(state, side, rand, extraData, renderType);
+    }
+
+    @Override
     public boolean useAmbientOcclusion() {
         return this.parent.useAmbientOcclusion();
+    }
+
+    @Override
+    public boolean useAmbientOcclusion(BlockState state, net.minecraft.client.renderer.RenderType renderType) {
+       return this.parent.useAmbientOcclusion(state, renderType);
     }
 
     @Override
