--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -94,25 +_,39 @@
         }
 
         @Override
+        public Function<Material, TextureAtlasSprite> getModelTextureGetter() {
+            return this.modelTextureGetter;
+        }
+
+        @Override
         public BakedModel bake(ResourceLocation p_252176_, ModelState p_249765_) {
+            return bake(p_252176_, p_249765_, this.modelTextureGetter);
+        }
+
+        @Override
+        public BakedModel bake(ResourceLocation p_252176_, ModelState p_249765_, Function<Material, TextureAtlasSprite> sprites) {
             ModelBakery.BakedCacheKey modelbakery$bakedcachekey = new ModelBakery.BakedCacheKey(p_252176_, p_249765_.getRotation(), p_249765_.isUvLocked());
             BakedModel bakedmodel = ModelBakery.this.bakedCache.get(modelbakery$bakedcachekey);
             if (bakedmodel != null) {
                 return bakedmodel;
             } else {
                 UnbakedModel unbakedmodel = this.getModel(p_252176_);
-                BakedModel bakedmodel1 = this.bakeUncached(unbakedmodel, p_249765_);
+                BakedModel bakedmodel1 = this.bakeUncached(unbakedmodel, p_249765_, sprites);
                 ModelBakery.this.bakedCache.put(modelbakery$bakedcachekey, bakedmodel1);
                 return bakedmodel1;
             }
         }
 
         BakedModel bakeUncached(UnbakedModel p_343761_, ModelState p_342939_) {
+            return bakeUncached(p_343761_, p_342939_, this.modelTextureGetter);
+        }
+
+        BakedModel bakeUncached(UnbakedModel p_343761_, ModelState p_342939_, Function<Material, TextureAtlasSprite> sprites) {
             if (p_343761_ instanceof BlockModel blockmodel && blockmodel.getRootModel() == SpecialModels.GENERATED_MARKER) {
-                return ModelBakery.ITEM_MODEL_GENERATOR.generateBlockModel(this.modelTextureGetter, blockmodel).bake(this.modelTextureGetter, p_342939_, false);
+                return ModelBakery.ITEM_MODEL_GENERATOR.generateBlockModel(sprites, blockmodel).bake(sprites, p_342939_, false);
             }
 
-            return p_343761_.bake(this, this.modelTextureGetter, p_342939_);
+            return p_343761_.bake(this, sprites, p_342939_);
         }
     }
 
