--- a/net/minecraft/data/recipes/RecipeOutput.java
+++ b/net/minecraft/data/recipes/RecipeOutput.java
@@ -7,7 +_,19 @@
 import net.minecraft.world.item.crafting.Recipe;
 
 public interface RecipeOutput {
-    void accept(ResourceKey<Recipe<?>> p_367162_, Recipe<?> p_360758_, @Nullable AdvancementHolder p_361155_);
+    default void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancement) {
+        if (advancement == null) {
+            accept(id, recipe, null, null);
+        } else {
+            var ops = registry().createSerializationContext(com.mojang.serialization.JsonOps.INSTANCE);
+            var json = Advancement.CODEC.encodeStart(ops, advancement.value()).getOrThrow(IllegalStateException::new);
+            accept(id, recipe, advancement.id(), json);
+        }
+    }
+
+    void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable net.minecraft.resources.ResourceLocation advancementId, @Nullable com.google.gson.JsonElement advancement);
+
+    net.minecraft.core.HolderLookup.Provider registry();
 
     Advancement.Builder advancement();
 
