--- a/net/minecraft/world/level/storage/loot/LootDataType.java
+++ b/net/minecraft/world/level/storage/loot/LootDataType.java
@@ -29,9 +_,12 @@
     }
 
     private static LootDataType.Validator<LootTable> createLootTableValidator() {
-        return (p_327551_, p_327552_, p_327553_) -> p_327553_.validate(
+        return (p_327551_, p_327552_, p_327553_) -> {
+            p_327553_.setLootTableId(p_327552_.location());
+            p_327553_.validate(
                 p_327551_.setContextKeySet(p_327553_.getParamSet()).enterElement("{" + p_327552_.registry() + "/" + p_327552_.location() + "}", p_327552_)
             );
+        };
     }
 
     @FunctionalInterface
