--- a/net/minecraft/world/entity/animal/Sheep.java
+++ b/net/minecraft/world/entity/animal/Sheep.java
@@ -161,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player p_29853_, InteractionHand p_29854_) {
         ItemStack itemstack = p_29853_.getItemInHand(p_29854_);
-        if (itemstack.is(Items.SHEARS)) {
+        if (false && itemstack.is(Items.SHEARS)) { // Forge: Moved to onSheared
             if (this.level() instanceof ServerLevel serverlevel && this.readyForShearing()) {
                 this.shear(serverlevel, SoundSource.PLAYERS, itemstack);
                 this.gameEvent(GameEvent.SHEAR, p_29853_);
@@ -177,12 +_,26 @@
 
     @Override
     public void shear(ServerLevel p_365174_, SoundSource p_29819_, ItemStack p_361058_) {
+        dropItems(p_365174_, shearInternal(p_365174_, p_29819_, p_361058_));
+    }
+
+    private java.util.List<ItemStack> shearInternal(ServerLevel p_365174_, SoundSource p_29819_, ItemStack p_361058_) {
+        var ret = new java.util.ArrayList<ItemStack>();
         p_365174_.playSound(null, this, SoundEvents.SHEEP_SHEAR, p_29819_, 1.0F, 1.0F);
         this.dropFromShearingLootTable(
             p_365174_,
             BuiltInLootTables.SHEAR_SHEEP,
             p_361058_,
             (p_359182_, p_359183_) -> {
+                ret.add(p_359183_);
+            }
+        );
+        this.setSheared(true);
+        return ret;
+    }
+
+    private void dropItems(ServerLevel p_359182_, java.util.Collection<ItemStack> items) {
+            for (var p_359183_ : items) {
                 for (int i = 0; i < p_359183_.getCount(); i++) {
                     ItemEntity itementity = this.spawnAtLocation(p_359182_, p_359183_.copyWithCount(1), 1.0F);
                     if (itementity != null) {
@@ -197,8 +_,14 @@
                     }
                 }
             }
-        );
-        this.setSheared(true);
+    }
+
+    @Override
+    public java.util.List<ItemStack> onSheared(@Nullable Player player, ItemStack item, Level world, BlockPos pos, int fortune) {
+        if (world instanceof ServerLevel server) {
+            return shearInternal(server, player == null ? SoundSource.BLOCKS : SoundSource.PLAYERS, item);
+        }
+        return java.util.Collections.emptyList();
     }
 
     @Override
