--- a/net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
+++ b/net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
@@ -11,7 +_,7 @@
     public static BehaviorControl<Piglin> create() {
         return BehaviorBuilder.create(
             p_259197_ -> p_259197_.group(p_259197_.absent(MemoryModuleType.ADMIRING_ITEM)).apply(p_259197_, p_259512_ -> (p_359309_, p_359310_, p_359311_) -> {
-                        if (!p_359310_.getOffhandItem().isEmpty() && !p_359310_.getOffhandItem().is(Items.SHIELD)) {
+                        if (!p_359310_.getOffhandItem().isEmpty() && !p_359310_.getOffhandItem().canPerformAction(net.minecraftforge.common.ToolActions.SHIELD_BLOCK)) {
                             PiglinAi.stopHoldingOffHandItem(p_359309_, p_359310_, true);
                             return true;
                         } else {
