--- a/net/minecraft/world/item/AnimalArmorItem.java
+++ b/net/minecraft/world/item/AnimalArmorItem.java
@@ -27,7 +_,7 @@
         return this.bodyType.breakingSound;
     }
 
-    public static enum BodyType {
+    public static enum BodyType implements net.minecraftforge.common.IExtensibleEnum {
         EQUESTRIAN(SoundEvents.ITEM_BREAK, EntityType.HORSE),
         CANINE(SoundEvents.WOLF_ARMOR_BREAK, EntityType.WOLF);
 
@@ -37,6 +_,10 @@
         private BodyType(final SoundEvent p_335661_, final EntityType<?>... p_364750_) {
             this.breakingSound = p_335661_;
             this.allowedEntities = HolderSet.direct(EntityType::builtInRegistryHolder, p_364750_);
+        }
+
+        public static BodyType create(String name, SoundEvent sound, EntityType<?>... types) {
+            throw new IllegalStateException("Enum not extended");
         }
     }
 }
