--- a/net/minecraft/world/item/BundleItem.java
+++ b/net/minecraft/world/item/BundleItem.java
@@ -98,6 +_,7 @@
 
     @Override
     public boolean overrideOtherStackedOnMe(ItemStack p_150742_, ItemStack p_150743_, Slot p_150744_, ClickAction p_150745_, Player p_150746_, SlotAccess p_150747_) {
+        if (p_150742_.getCount() != 1) return false;
         if (p_150745_ == ClickAction.PRIMARY && p_150743_.isEmpty()) {
             toggleSelectedItem(p_150742_, -1);
             return false;
