--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -2800,7 +_,7 @@
     );
     public static final Block QUARTZ_STAIRS = registerLegacyStair("quartz_stairs", QUARTZ_BLOCK);
     public static final Block ACTIVATOR_RAIL = register(
-        "activator_rail", PoweredRailBlock::new, BlockBehaviour.Properties.of().noCollission().strength(0.7F).sound(SoundType.METAL)
+        "activator_rail", prop -> new PoweredRailBlock(prop, false), BlockBehaviour.Properties.of().noCollission().strength(0.7F).sound(SoundType.METAL)
     );
     public static final Block DROPPER = register(
         "dropper",
@@ -6552,14 +_,5 @@
 
     private static Block register(String p_50796_, BlockBehaviour.Properties p_362760_) {
         return register(p_50796_, Block::new, p_362760_);
-    }
-
-    static {
-        for (Block block : BuiltInRegistries.BLOCK) {
-            for (BlockState blockstate : block.getStateDefinition().getPossibleStates()) {
-                Block.BLOCK_STATE_REGISTRY.add(blockstate);
-                blockstate.initCache();
-            }
-        }
     }
 }
