--- a/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
@@ -98,6 +_,7 @@
 
             int j = mapid != null ? p_361692_.rotation % 4 * 2 : p_361692_.rotation;
             p_115061_.mulPose(Axis.ZP.rotationDegrees((float)j * 360.0F / 8.0F));
+            if (!net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderItemInFrame(p_361692_, this, p_115061_, p_115062_, p_115063_)) {
             if (mapid != null) {
                 p_115061_.mulPose(Axis.ZP.rotationDegrees(180.0F));
                 float f2 = 0.0078125F;
@@ -110,6 +_,7 @@
                 int k = this.getLightVal(p_361692_.isGlowFrame, 15728880, p_115063_);
                 p_115061_.scale(0.5F, 0.5F, 0.5F);
                 this.itemRenderer.render(itemstack, ItemDisplayContext.FIXED, false, p_115061_, p_115062_, k, OverlayTexture.NO_OVERLAY, p_361692_.itemModel);
+            }
             }
         }
 
