--- a/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
@@ -69,6 +_,7 @@
     }
 
     public void render(S p_364280_, PoseStack p_115311_, MultiBufferSource p_115312_, int p_115313_) {
+        if (net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderLivingPre(p_364280_, this, p_115311_, p_115312_, p_115313_)) return;
         p_115311_.pushPose();
         if (p_364280_.hasPose(Pose.SLEEPING)) {
             Direction direction = p_364280_.bedOrientation;
@@ -104,6 +_,7 @@
 
         p_115311_.popPose();
         super.render(p_364280_, p_115311_, p_115312_, p_115313_);
+        net.minecraftforge.client.event.ForgeEventFactoryClient.onRenderLivingPost(p_364280_, this, p_115311_, p_115312_, p_115313_);
     }
 
     protected boolean shouldRenderLayers(S p_360804_) {
@@ -284,7 +_,7 @@
 
         p_363057_.isFullyFrozen = p_368665_.isFullyFrozen();
         p_363057_.isBaby = p_368665_.isBaby();
-        p_363057_.isInWater = p_368665_.isInWater();
+        p_363057_.isInWater = p_368665_.isInWater() || p_368665_.isInFluidType((fluidType, height) -> p_368665_.canSwimInFluidType(fluidType));
         p_363057_.isAutoSpinAttack = p_368665_.isAutoSpinAttack();
         p_363057_.hasRedOverlay = p_368665_.hurtTime > 0 || p_368665_.deathTime > 0;
         ItemStack itemstack1 = p_368665_.getItemBySlot(EquipmentSlot.HEAD);
