--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -58,7 +_,7 @@
         this.updateRotation();
         this.applyEffectsFromBlocks();
         super.tick();
-        if (hitresult.getType() != HitResult.Type.MISS && this.isAlive()) {
+        if (hitresult.getType() != HitResult.Type.MISS && this.isAlive() && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
             this.hitTargetOrDeflectSelf(hitresult);
         }
     }
