--- a/net/minecraft/client/renderer/block/model/MultiVariant.java
+++ b/net/minecraft/client/renderer/block/model/MultiVariant.java
@@ -50,7 +_,7 @@
             SimpleWeightedRandomList.Builder<BakedModel> builder = SimpleWeightedRandomList.builder();
 
             for (Variant variant : this.variants) {
-                BakedModel bakedmodel = p_249016_.bake(variant.getModelLocation(), variant);
+                BakedModel bakedmodel = p_249016_.bake(variant.getModelLocation(), variant, p_111851_);
                 builder.add(bakedmodel, variant.getWeight());
             }
 
