--- a/com/mojang/blaze3d/systems/RenderSystem.java
+++ b/com/mojang/blaze3d/systems/RenderSystem.java
@@ -196,6 +_,11 @@
         GlStateManager._depthMask(p_69459_);
     }
 
+    public static boolean isBlendEnabled() {
+        assertOnRenderThread();
+        return GlStateManager._isBlendEnabled();
+    }
+
     public static void enableBlend() {
         assertOnRenderThread();
         GlStateManager._enableBlend();
