--- a/net/minecraft/util/context/ContextMap.java
+++ b/net/minecraft/util/context/ContextMap.java
@@ -39,6 +_,10 @@
         return (T)this.params.getOrDefault(p_368195_, p_368521_);
     }
 
+    public Set<ContextKey<?>> keySet() {
+        return java.util.Collections.unmodifiableSet(this.params.keySet());
+    }
+
     public static class Builder {
         private final Map<ContextKey<?>, Object> params = new IdentityHashMap<>();
 
@@ -83,6 +_,10 @@
                     return new ContextMap(this.params);
                 }
             }
+        }
+
+        public ContextMap create() {
+            return new ContextMap(this.params);
         }
     }
 }
