--- a/net/minecraft/client/renderer/entity/state/PlayerRenderState.java
+++ b/net/minecraft/client/renderer/entity/state/PlayerRenderState.java
@@ -48,6 +_,8 @@
 
     @OnlyIn(Dist.CLIENT)
     public static class HandState {
+        @Nullable
+        public net.minecraft.client.model.HumanoidModel.ArmPose pose;
         public boolean isEmpty = true;
         @Nullable
         public ItemUseAnimation useAnimation;
